### PR TITLE
Refactor for BigQueryPartition class

### PR DESCRIPTION
**Pull request by prodriguezdefino**
In preparation for the unbounded source implementation. Test coverage for partition related methods was improved as well.

### CODE REVIEW

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/services/TablePartitionInfo.java [hunk 1]
```diff
@@ -18,24 +18,34 @@
 
 import org.apache.flink.annotation.Internal;
 
+import org.apache.flink.shaded.guava30.com.google.common.collect.Lists;
```
Comment (vishalkarve15) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/services/TablePartitionInfo.java:21]: We can avoid this guava usage. Are we addressing this later?

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/common/utils/BigQueryPartition.java [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.TimeZone;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/** Utility class to handle the BigQuery partition conversions to Flink types and structures. */
+@Internal
+public class BigQueryPartition {
```
Comment (vishalkarve15) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/common/utils/BigQueryPartition.java:50]: This should be named BigQueryPartitionUtils

Comment (vishalkarve15) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/common/utils/BigQueryPartition.java:50]: This file is hard to review since it also has some changes. Is it possible to move the file in a separate PR? Or move it after the review?

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/common/utils/BigQueryPartitionTest.java [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+import com.google.cloud.flink.bigquery.services.TablePartitionInfo;
+import org.assertj.core.api.Assertions;
+import org.junit.Test;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+
+/** */
+public class BigQueryPartitionTest {
```
Comment (vishalkarve15) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/common/utils/BigQueryPartitionTest.java:34]: This file is also hard to review due to the move. can we move it at the end after reviewing is complete?

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/services/TablePartitionInfo.java (hunk 1 above)
Comment (prodriguezdefino) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/services/TablePartitionInfo.java:21]: Yea PR #60

Comment (jayehwhyehentee) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/services/TablePartitionInfo.java:21]: Yes.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/common/utils/BigQueryPartition.java (hunk 2 above)
Comment (jayehwhyehentee) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/common/utils/BigQueryPartition.java]: Reverted file relocation. Just refactoring here.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartition.java [hunk 4]
```diff
@@ -19,26 +19,83 @@
...
+import java.util.function.Function;
 import java.util.stream.Collectors;
 
 /** Utility class to handle the BigQuery partition conversions to Flink types and structures. */
 @Internal
 public class BigQueryPartition {
 
+    static final Integer HOUR_SECONDS = 3600;
+    static final Integer DAY_SECONDS = 86400;
+    static final Integer MONTH_SECONDS = 2629746;
```
Comment (vishalkarve15) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartition.java:54]: How do we get this number?

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartition.java [hunk 5]
```diff
@@ -19,26 +19,83 @@
...
+    static final Integer YEAR_SECONDS = 31536000;
```
Comment (vishalkarve15) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartition.java:55]: Will this be different for a leap year?

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java [hunk 6]
```diff
@@ -179,166 +181,311 @@ public void testWrongNumericPartition() {
...
 
-    @Test(expected = IllegalArgumentException.class)
-    public void testDateFormatHourPartitionRestriction() {
-        TablePartitionInfo info =
-                new TablePartitionInfo(
-                        "event", BigQueryPartition.PartitionType.HOUR, StandardSQLTypeName.DATE);
-        BigQueryPartition.formatPartitionRestrictionBasedOnInfo(
-                Optional.of(info), "event", "2010-10-10");
+    @Test
+    public void testCheckPartitionNotCompletedYear() {
```
Comment (vishalkarve15) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java:309]: We should add edge case tests for leap year.

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java [hunk 7]
```diff
@@ -179,166 +181,311 @@ public void testWrongNumericPartition() {
...
     @Test
-    public void testDateFormatMonthPartitionRestriction() {
-        TablePartitionInfo info =
-                new TablePartitionInfo(
-                        "event", BigQueryPartition.PartitionType.MONTH, StandardSQLTypeName.DATE);
-        Assertions.assertThat(
-                        BigQueryPartition.formatPartitionRestrictionBasedOnInfo(
-                                Optional.of(info), "event", "2010-10-10"))
-                .isEqualTo("event BETWEEN '2010-10' AND '2010-11'");
+    public void testCheckPartitionCompletedMonth() {
```
Comment (vishalkarve15) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java:271]: We should add edge case tests for February as leap month, non-leap month, a 30 day month, a 31 day month.

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java [hunk 8]
```diff
@@ -179,166 +181,311 @@ public void testWrongNumericPartition() {
...
+    public void testCheckPartitionNotCompletedYear() {
+        PartitionIdWithInfo partitionWithInfo =
+                new PartitionIdWithInfo(
+                        "2023",
+                        new TablePartitionInfo(
+                                "temporal",
+                                BigQueryPartition.PartitionType.YEAR,
+                                StandardSQLTypeName.DATE,
+                                // partition watermark is now
+                                Instant.now()));
```
Comment (vishalkarve15) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java:318]: oldestEntryTime can be upto 90 mins ago. Let's add tests for those cases.

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java [hunk 9]
```diff
@@ -179,166 +181,311 @@ public void testWrongNumericPartition() {
...
     @Test
-    public void testDateFormatDayPartitionRestriction() {
-        TablePartitionInfo info =
-                new TablePartitionInfo(
-                        "event", BigQueryPartition.PartitionType.DAY, StandardSQLTypeName.DATE);
-        Assertions.assertThat(
-                        BigQueryPartition.formatPartitionRestrictionBasedOnInfo(
-                                Optional.of(info), "event", "2010-10-10"))
-                .isEqualTo("event BETWEEN '2010-10-10' AND '2010-10-11'");
+    public void testCheckPartitionCompletedDay() {
```
Comment (vishalkarve15) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java:252]: Let's add edge cases near end of day, end of hour etc.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartition.java (hunk 5 above)
Comment (jayehwhyehentee) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartition.java]: Addressed

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java (hunk 6 above)
Comment (jayehwhyehentee) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java]: Done

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java (hunk 7 above)
Comment (jayehwhyehentee) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java:307]: Done

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java (hunk 9 above)
Comment (jayehwhyehentee) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java:271]: Done

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java (hunk 7 above)
Comment (vishalkarve15) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java:307]: Please add not completed 31 day month case. LGTM otherwise.

Comment (jayehwhyehentee) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/restrictions/BigQueryPartitionTest.java:307]: Done

**Review by vishalkarve15 — APPROVED** [review]
